internal/provider/general_api_models.go

```diff
@@ -16,3 +31,227 @@ type SummaryAPIModel struct {
+	d.Organization = types.Int64Value(apiWorkflowJobTemplate.Organization)
+	d.OrganizationName = ParseStringValue(apiWorkflowJobTemplate.SummaryFields.Organization.Name)
+	d.Url = ParseStringValue(apiWorkflowJobTemplate.Url)
+	d.NamedUrl = ParseStringValue(apiWorkflowJobTemplate.Related.NamedUrl)
+	d.Name = ParseStringValue(apiWorkflowJobTemplate.Name)
+	d.Description = ParseStringValue(apiWorkflowJobTemplate.Description)
+	d.Variables = ParseAAPCustomStringValue(apiWorkflowJobTemplate.Variables)
+
+	return diags
+}
+
+type StringDescriptions struct {
+	ApiEntitySlug         string
+	DescriptiveEntityName string
+	MetadataEntitySlug    string
+}
+
+// A struct to represent a base DataSource object, with a client and the slug name of
+// the API entity.
+type BaseDataSource struct {
+	client ProviderHTTPClient
+	StringDescriptions
+}
+
+// Constructs a new BaseDataSource object provided with a client instance (usually
+// initialized to nil, it will be later configured calling the Configure function)
+// and an apiEntitySlug string indicating the entity path name to consult the API.
+func NewBaseDataSource(client ProviderHTTPClient, stringDescriptions StringDescriptions) *BaseDataSource {
+	return &BaseDataSource{
+		client:             client,
+		StringDescriptions: stringDescriptions,
+	}
+}
+
+// Read refreshes the Terraform state with the latest data.
+func (d *BaseDataSource) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	var state BaseDataSourceModel
+	var diags diag.Diagnostics
+
+	// Read Terraform configuration data into the model
+	resp.Diagnostics.Append(req.Config.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	uri := path.Join(d.client.getApiEndpoint(), d.ApiEntitySlug)
+	resourceURL, err := ReturnAAPNamedURL(state.Id, state.Name, state.OrganizationName, uri)
+	if err != nil {
+		resp.Diagnostics.AddError("Minimal Data Not Supplied", "Expected either [id] or [name + organization_name] pair")
+		return
+	}
+
+	readResponseBody, diags := d.client.Get(resourceURL)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	diags = state.ParseHttpResponse(readResponseBody)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	// Set state
+	// Save updated data into Terraform state
+	resp.Diagnostics.Append(resp.State.Set(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+}
+
+// Configure adds the provider configured client to the data source.
+func (d *BaseDataSource) Configure(_ context.Context, req datasource.ConfigureRequest, resp *datasource.ConfigureResponse) {
+	if req.ProviderData == nil {
+		return
+	}
+
+	client, ok := req.ProviderData.(*AAPClient)
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Data Source Configure Type",
+			fmt.Sprintf("Expected *AAPClient, got: %T. Please report this issue to the provider developers.", req.ProviderData),
```

Comment:
We dont know what `req.ProviderData` is but we attempt to print it, Does this open us up for a log injection attack? Should we attempt to `Sanitize` this `req.ProviderData` before printing, I think we should

Reply:
Also I just found out that we shouldnt be using fmt.Sprintf for logging, link: https://developer.hashicorp.com/terraform/plugin/log/writing